### PR TITLE
feat(testing): module for testing utilities

### PR DESCRIPTION
**Pull request by WonderPanda**
re [AB#20](https://dev.azure.com/golevelup/2e65042c-ccd9-4a3f-987f-075ebfc5a042/_workitems/edit/20)

### CODE REVIEW

**Comment by jmcdo29** [comment]
Man it's amazing how small that it but how well it works! I'd say it'd be a good idea to see how it works with ArgumentHost and maybe some random class, just to be extra sure, but that looks awesome for a mock utility

**Comment by WonderPanda** [comment]
> Man it's amazing how small that it but how well it works! I'd say it'd be a good idea to see how it works with ArgumentHost and maybe some random class, just to be extra sure, but that looks awesome for a mock utility

Did you see this test, I think it really shows off how powerful the function is and includes `ArgumentHost`:

```
it('should allow for mock implementation on automocked properties', () => {
      const executionContextMock = makeMock<ExecutionContext>();
      const httpArgsHost = makeMock<HttpArgumentsHost>({
        getRequest: () => request
      });

       executionContextMock.switchToHttp.mockImplementation(() => httpArgsHost);

       const result = executionContextMock.switchToHttp().getRequest();
      expect(result).toBe(request);
      expect(httpArgsHost.getRequest).toBeCalledTimes(1);
    });
```

**Comment by jmcdo29** [comment]
I totally missed that treat! Nice! 

Only last question is does it okay well work GraphQL, but that can be an enhancement if not

**Comment by WonderPanda** [comment]
> I totally missed that test! Nice!
> 
> Only last question is does it okay well work GraphQL, but that can be an enhancement if not

It's pretty general purpose, right now it would work with anything that is a JS object. Gonna take a look to see if it makes sense to have a separate function when mocking straight functions but I'll probably publish 0.0.1 shortly and then open a PR against your testing Repo

**Comment by jmcdo29** [comment]
Let me know and I'll test it against my main server as well, as it had a GraphQL guard. I should make a complex GraphQL sample for the testing repo though

File: packages/testing/src/mocks.ts [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+  const cache = new Map<string | number | symbol, any>();
+
+  const proxy = new Proxy(partial, {
+    get: (obj, prop) => {
+      if (cache.has(prop)) {
+        return cache.get(prop);
+      }
+
+      const anyObj = obj as any;
+      const checkProp = anyObj[prop];
```
Comment (jmcdo29) [inline packages/testing/src/mocks.ts]: Any chance you'd want to change this to declare mockedProp before the if/else, set it in the if/else and then cache and return it after the if/else? Just so you're reducing the number of places you do the same thing.
```ts
let mockedProp;
if (checkProp) {
  mockedProp = typeof checkProp === "function" ? jest.fn(checkProp) : checkProp;
} else {
  mockedProp = makeRecursiveMockProxy();
}
cache.set(prop, mockedProp);
return mockedProp;
```

Comment (jmcdo29) [reply packages/testing/src/mocks.ts]: Hell, you could technically make it a (slightly complex) one line ternary, though that may be a bit much
```ts
const mockedProp = checkProp 
    ? typeof checkedProp === 'function' 
        ? jest.fn()
        : checkedProp
    : makeRecursiveProxy();
```

Comment (WonderPanda) [reply packages/testing/src/mocks.ts]: Good call I like the way this looks a lot better

**Comment by jmcdo29** [comment]
I'm excited to see this go public! Really think it'll be good for the community

**Pull request merged**